### PR TITLE
feat: solidity style guide

### PR DESCRIPTION
**Pull request by nksazonov**

### CODE REVIEW

File: contracts/solidity-style-guide.md [hunk 1]
```diff
@@ -0,0 +1,566 @@
...
+
+NO:
+
+```solidity
+pow(x, y, v);
+```
+
+##### C. Prefer named parameters in mapping types
+
+Explicit naming parameters in mapping types is helpful for clarity, especially when nesting is used.
```
Comment (maxpushka) [inline contracts/solidity-style-guide.md:158]:
```suggestion
Explicitly naming parameters in mapping types is helpful for clarity, especially when nesting is used.
```

**Review by maxpushka — CHANGES_REQUESTED** [review]

File: contracts/solidity-style-guide.md [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+
+#### 1. Errors
+
+##### A. Prefer custom errors
+
+Custom errors are in some cases more gas efficient and allow passing useful information.
+
+##### B. Custom error names should be CapWords style
+
+For example, `InsufficientBalance`.
```
[the remark this answers is not in the document]
Comment (nksazonov) [reply contracts/solidity-style-guide.md]: 7ab77dbd87c0e4f2eaeed3e9d85133c579066b06

File: contracts/solidity-style-guide.md [hunk 3]
```diff
@@ -0,0 +1,566 @@
...
+
+One of the motivations for this rule is that it is a helpful visual clue.
+
+> Leading underscores allow you to immediately recognize the intent of such functions...
+
+We agree that a leading underscore is a useful visual clue, and this is why we oppose using them for internal library functions that can be called from other contracts. Visually, it looks wrong.
+
+```solidity
+Library._function()
+```
```
[the remark this answers is not in the document]
Comment (nksazonov) [reply contracts/solidity-style-guide.md:25]: 7ab77dbd87c0e4f2eaeed3e9d85133c579066b06

File: contracts/solidity-style-guide.md [hunk 4]
```diff
@@ -0,0 +1,566 @@
...
+
+or
+
+```solidity
+using Library for bytes
+bytes._function()
+```
```
[the remark this answers is not in the document]
Comment (nksazonov) [reply contracts/solidity-style-guide.md:32]: 7ab77dbd87c0e4f2eaeed3e9d85133c579066b06

File: contracts/solidity-style-guide.md [hunk 5]
```diff
@@ -0,0 +1,566 @@
...
+
+##### B. Avoid using assembly
+
+Assembly code is hard to read and audit. We should avoid it unless the gas savings are very consequential, e.g. > 25%.
+
+#### 4. Versioning
+
+##### A. Avoid unnecessary version Pragma constraints
+
+While the main contracts we deploy should specify a single Solidity version, all supporting contracts and libraries should have as open a Pragma as possible. A good rule of thumb is to the next major version. For example
```
[the remark this answers is not in the document]
Comment (nksazonov) [reply contracts/solidity-style-guide.md]: 7ab77dbd87c0e4f2eaeed3e9d85133c579066b06

File: contracts/solidity-style-guide.md [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+```
+
+or
+
+```solidity
+using Library for bytes
+bytes._function()
+```
+
+Note, we cannot remedy this by insisting on the use public functions. Whether a library functions are internal or external has important implications. From the [Solidity documentation](https://docs.soliditylang.org/en/latest/contracts.html#libraries)
```
[the remark this answers is not in the document]
Comment (nksazonov) [reply contracts/solidity-style-guide.md]: 7ab77dbd87c0e4f2eaeed3e9d85133c579066b06

File: contracts/solidity-style-guide.md (hunk 1 above)
Comment (nksazonov) [reply contracts/solidity-style-guide.md]: 7ab77dbd87c0e4f2eaeed3e9d85133c579066b06

File: contracts/solidity-style-guide.md [hunk 7]
```diff
@@ -0,0 +1,566 @@
+# Solidity Style Guide
+
+This style guide is a set of rules and conventions for writing Solidity code. It is based on the [Solidity Style Guide](https://docs.soliditylang.org/en/latest/style-guide.html), [Coinbase Style Guide](https://github.com/coinbase/solidity-style-guide/blob/7e29d890be81b7df9beefb38f04ff212cc32fb6c/README.md) and includes some additional rules and conventions specific to Yellow Network.
```
[the remark this answers is not in the document]
Comment (nksazonov) [reply contracts/solidity-style-guide.md:3]: > Use rule structure similar to [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html):

In the most controversial and opinionated cases this structure is a great approach, whereas in general the simple definition or explanation should suffice.

Nevertheless, we can change this if we see the need to it.

[the remark this answers is not in the document]
Comment (nksazonov) [reply contracts/solidity-style-guide.md:3]: > Add some common gas usage optimizations

Indeed, I had some thoughts of adding gas optimization technics to this document, however later realized that alongside the security technics, they should be outlined in a separate document.

The Style Guide declares the syntax and semantics of writing contracts, whereas gas optimizations and security technics are more complex terms.

Still, great thanks for reminding about this topic!

[the remark this answers is not in the document]
Comment (nksazonov) [reply contracts/solidity-style-guide.md:3]: > Add some common gas usage optimizations

@maxpushka 
I have added a [Solidity Development Practices](https://github.com/layer-3/clearsync/blob/f7dedf31dc9858d59c44366423f542eb8e834829/contracts/solidity-development-practices.md) document to store optimizations and security technics there.

**Review by dpatsora — APPROVED** [review]
Nikita, thank you for such comprehensive guides!!

**Pull request merged**